boxoffice/models/line_item.py

```diff
@@ -1,47 +1,212 @@
+                # if the discount policy's upper limit hasn't been reached,and if the line
+                # item hasn't been assigned a discount or if the line item's assigned discount is lesser
+                # than the current discount, assign the current discount to the line item
+
+                discount_coupon_id = coupon.id if coupon else None
+                discounted_line_items.append(LineItem.make_ntuple(item_id=line_item.item_id,
+                    base_amount=line_item.base_amount,
+                    discount_policy_id=discount_policy.id,
+                    discount_coupon_id=discount_coupon_id,
+                    discounted_amount=discounted_amount))
+
+                applied_to_count += 1
+                if discount_policy.item_quantity_max and applied_to_count == discount_policy.item_quantity_max:
+                    # If the discount policy has a upper limit and the upper limit on the discount's allowed item quantity
+                    # is reached, stop iterating through line items.
+                    should_apply_discount = False
+            else:
+                # Current discount is not applicable, copy over the line item as it is.
+                discounted_line_items.append(line_item)
+        return discounted_line_items
+
+    def apply_combo_discount(self, discounts, line_items):
+        """
+        Applies multiple discounts to a list of line items recursively.
+        """
+        if len(discounts) == 0:
+            return line_items
+        if len(discounts) == 1:
+            return self.apply_discount(discounts[0], line_items, combo=True)
+        return self.apply_combo_discount([discounts[0]], self.apply_combo_discount(discounts[1:], line_items))
+
+    def apply_max_discount(self, discounts, line_items):
+        """
+        Fetches the various discount combinations and applies the discount policy
+        combination that results in the maximum dicount for the given list of
+        line items.
+        """
+        discounts.extend(self.get_combos(discounts, len(line_items)))
+        discounted_line_items_list = []
+
+        for discount in discounts:
+            if isinstance(discount[0], tuple):
+                # Combo discount
+                discounted_line_items_list.append(self.apply_combo_discount(discount, line_items))
+            else:
+                discounted_line_items_list.append(self.apply_discount(discount, line_items))
+        return max(discounted_line_items_list,
+            key=lambda line_item_list: sum([line_item.discounted_amount for line_item in line_item_list]))
+
+    def get_combos(self, discounts, qty):
+        """
+        Returns the various valid discount combinations given a list of discount policies
+        """
+        valid_combos = []
+        # Get all possible valid combinations of discounts in groups of 2..n,
+        # where n is the number of discounts
+        for n in range(2, len(discounts) + 1):
```

Comment:
Added an additional check; it will return early with an empty array if `len(discounts) < 2`.